### ISSUE TITLE
Excessive print output from mermaid example in CLI on Windows

### ISSUE THREAD

**Issue by nathanjmcdougall** [ben-denham/labtech#16]
Similar to #14, running the mermaid example from the README from the CLI (rather than an interactive IPython session) seems to spam the console with 12 times. As of ef08cfee7f1db905b04c2fa9e2ff7c15d4ce03d4, this is no longer `<IPython.core.display.Markdown object>` but instead, a text representation of the diagram.

**Comment by nathanjmcdougall** [comment]
A few things:

1. I was running the code from a file rather than in a terminal. Actually I was just running the test file `dependents_and_mermaid.py`.
2. I wonder whether it is `tqdm` related. I have noticed that sometimes when using `tqdm` I get issues where output is repeated if I move/resize the window while it is running. However this issue is consistent so I am not sure.
3. When I try to run the code directly in the terminal as per your screenshot above I actually get spawn process errors. I don't have time right now to look into this further but it seems unrelated, but worth investigating further.
4. I am using Powershell from within VS Code. Nothing else unusual I don't think.

**Comment by ben-denham** [comment]
Thanks @nathanjmcdougall 
1. Good to know. I wouldn't expect this to really make a difference though.
2. Interesting idea, I've just added https://github.com/ben-denham/labtech/commit/82b7eeccd78d75d776fc7aae2345e01d6aebaa26 to see if that makes terminal resizing any better (though it won't fix already-printed lines)
3. Would be keen to see those errors, maybe paste a screenshot in a separate issue?
4. Could you please see if you can get the same issue running straight in Powershell - not in VS Code? Just to try and isolate the issue.

**Comment by nathanjmcdougall** [comment]
Issue 4 is reproduced within Powershell only, without using VS Code:

![image](https://github.com/ben-denham/labtech/assets/18602289/0b073870-20ee-4d7c-b948-751a70dbd8b6)

<details>

```
PS C:\Users\namc> cd Repositories/labtech
PS C:\Users\namc\Repositories\labtech> .venv\Scripts\activate
(.venv) PS C:\Users\namc\Repositories\labtech> poetry run python tests\integration\readme\dependents_and_mermaid.py
classDiagram
    direction BT

    class DependentTask
    DependentTask : SlowTask slow_task
    DependentTask : int multiplier
    DependentTask : run() int

    class SlowTask
    SlowTask : int base
    SlowTask : run() int


    DependentTask <-- SlowTask: slow_task
DependentTask: 0/10 [00:00, ?tasks/s] |                                                                                                                                                       |   0% [? remaining]classDiagram
    direction BT

    class DependentTask
    DependentTask : SlowTask slow_task
    DependentTask : int multiplier
    DependentTask : run() int

    class SlowTask
    SlowTask : int base
    SlowTask : run() int


    DependentTask <-- SlowTask: slow_task
classDiagram
    direction BTs at 20:46:00. Up to top 10 by start_time:

    class DependentTask
    DependentTask : SlowTask slow_task
    DependentTask : int multiplier
    DependentTask : run() int

    class SlowTask
    SlowTask : int base
    SlowTask : run() int


    DependentTask <-- SlowTask: slow_task
classDiagram
    direction BT

    class DependentTask
    DependentTask : SlowTask slow_task
    DependentTask : int multiplier
    DependentTask : run() int

    class SlowTask
    SlowTask : int base
    SlowTask : run() int


    DependentTask <-- SlowTask: slow_task
classDiagram
    direction BT

    class DependentTask
    DependentTask : SlowTask slow_task
    DependentTask : int multiplier
    DependentTask : run() int

    class SlowTask
    SlowTask : int base
    SlowTask : run() int


    DependentTask <-- SlowTask: slow_task
DependentTask: 1/10 [00:01,  1.84s/tasks] |██████████████▎                                                                                                                                |  10% [00:16 remaining]classDiagram
    direction BT

    class DependentTask
    DependentTask : SlowTask slow_task
    DependentTask : int multiplier
    DependentTask : run() int

    class SlowTask
    SlowTask : int base
    SlowTask : run() int


    DependentTask <-- SlowTask: slow_task
classDiagram
    direction BT

    class DependentTask
    DependentTask : SlowTask slow_task
    DependentTask : int multiplier
    DependentTask : run() int

    class SlowTask
    SlowTask : int base
    SlowTask : run() int


    DependentTask <-- SlowTask: slow_taskclassDiagram
    direction BT

    class DependentTask
    DependentTask : SlowTask slow_task
    DependentTask : int multiplier
    DependentTask : run() int

    class SlowTask
    SlowTask : int base
    SlowTask : run() int


    DependentTask <-- SlowTask: slow_task

classDiagram
    direction BT

    class DependentTask
    DependentTask : SlowTask slow_task
    DependentTask : int multiplier
    DependentTask : run() int

    class SlowTask
    SlowTask : int base
    SlowTask : run() int


    DependentTask <-- SlowTask: slow_task
classDiagram
    direction BT

    class DependentTask
    DependentTask : SlowTask slow_task
    DependentTask : int multiplier
    DependentTask : run() int

    class SlowTask
    SlowTask : int base
    SlowTask : run() int


    DependentTask <-- SlowTask: slow_task
classDiagram
    direction BT

    class DependentTask
    DependentTask : SlowTask slow_task
    DependentTask : int multiplier
    DependentTask : run() int

    class SlowTask
    SlowTask : int base
    SlowTask : run() int


    DependentTask <-- SlowTask: slow_task
classDiagram
0 active tasks as at 20:46:01. Up to top 10 by start_time:










[0, 1764, 3528, 5292, 7056, 8820, 10584, 12348, 14112, 15876]
classDiagramtTask <-- SlowTask: slow_task
    direction BT0/10 [00:02,  4.67tasks/s] |██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100% [00:00 remaining]
0 active tasks as at 20:46:01. Up to top 10 by start_time:
    class DependentTask:46:01. Up to top 10 by start_time:
    DependentTask : SlowTask slow_task
    DependentTask : int multiplier
    DependentTask : run() int

    class SlowTask
    SlowTask : int base
    SlowTask : run() int


    DependentTask <-- SlowTask: slow_task
```
</details>

**Comment by nathanjmcdougall** [comment]
Regarding point 3, see #25.

**Comment by ben-denham** [comment]
Aha, I was able to reproduce the issue by running the specific test as a module just as you did.

The cause of the issue is that the diagram-printing lines are at the top-level of module and unguarded by `if __name__ == '__main__':`, which means they will be re-executed whenever the module is loaded. With the default multiprocessing `start_method` of `'spawn'` on Windows, the module will be reloaded at the start of each subprocess (skipping over anything guarded by `if __name__ == '__main__':`), printing the diagram from each subprocess.

I've fixed this by moving the diagram printing under `if __name__ == '__main__':` in https://github.com/ben-denham/labtech/commit/ddb677cc09f76d9b55ebede04cd0bb1d1a27ea22

**Comment by nathanjmcdougall** [comment]
Okay, but then won't those lines be untested/unexecuted in the test suite?
Might be better to add those lines to the README itself but exclude then
from the integration tests.

**Comment by ben-denham** [comment]
They'll still be executed, but only once in the main process instead of repeatedly in each subprocess. I've updated the test to check for exactly one diagram in the output: https://github.com/ben-denham/labtech/commit/ddb677cc09f76d9b55ebede04cd0bb1d1a27ea22#diff-eb0ef88a399d99af6bfcec93c9fdbb2e4656d6f3cc05150ef6bf4fe736d721a6R35

**Comment by nathanjmcdougall** [comment]
Ah, great!
